### PR TITLE
CSS animations in SVG do not repaint when SVG is zoomed

### PR DESCRIPTION
**Pull request by Ahmad-S792**
<pre>
CSS animations in SVG do not repaint when SVG is zoomed
<a href="https://bugs.webkit.org/show_bug.cgi?id=199364">https://bugs.webkit.org/show_bug.cgi?id=199364</a>
rdar://problem/52459928

Reviewed by NOBODY (OOPS!).

Partial Merge - <a href="https://src.chromium.org/viewvc/blink?view=revision&revision=174874">https://src.chromium.org/viewvc/blink?view=revision&revision=174874</a>

This patch changes the coordinate system used by RenderImage for invalidation
rects to account for zoom. When an SVG image issues invalidations to its
ImageResource clients, the invalidation rect does not contain zoom. This is
due to how SVG images work internally: images are laid out without zoom so
that relative sizes are resolved correctly; as a result the invalidation
rects do not include zoom. This patch simply removes the zoom from the image
size in RenderImage::repaintOrMarkForLayout and clarifies the comment.

This change should not affect bitmap image invalidation because bitmaps (even
animated gifs) do not use invalidation rects.

From local build testing, I have verified that this does indeed fix an issue
described on test case from the bug.

* Source/WebCore/rendering/RenderImage.cpp:
(RenderImage::repaintOrMarkForLayout): Update Comment and simplify logic
to remove for 'zoom' factor
* LayoutTests/svg/repaint/resources/animate-center.svg: Add Test Case Resource
* LayoutTests/svg/repaint/image-animation-with-zoom.html: Add Test Case
* LayoutTests/svg/repaint/image-animation-with-zoom-expected.txt: Add Test Case Expectation
</pre>

https://github.com/WebKit/WebKit/commit/84bc61de1e17029059fff86520fea6f1b0ab547b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105292 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14375 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38183 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114552 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174745 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109201 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15544 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5298 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97609 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114471 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111048 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12030 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95018 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39514 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93896 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26652 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81179 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7703 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28012 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7802 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4589 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13848 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47556 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9590 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Comment by Ahmad-S792** [comment]
NOTE - I didn't took the test case from Blink patch because in [PR4727](https://github.com/WebKit/WebKit/pull/4727), @shallawa mentioned that this will lead to flaky issue due to timeout of 220. Although, in previous attempt, @shallawa did local testing and the patch was not fixing the issue but upon later local testing, I verified that this does fix the issue and it might be due to removing CSS border-width zoom quirk in this [commit](https://github.com/WebKit/WebKit/commit/e7f78cb4cd8136f0b3c6e0c02b36009022514452). Hence, I thought to do PR again after local testing without test-case.

**Comment by Ahmad-S792** [comment]
NOTE - 'windows' build failure error seems to suggest 'JavaScriptCoreSharedScripts.rule' exited with code 1.', so nothing seems to suggest that this could be because of this change.

**Comment by Ahmad-S792** [comment]
@smfr - I took test from Chromium patch and tried to apply two tricks to make it less flaky fo future: 1) Small Size of Image 2) Shorter set-timeout. Can you have suggest any other?

**Comment by Ahmad-S792** [comment]
> @smfr - I took test from Chromium patch and tried to apply two tricks to make it less flaky fo future: 1) Small Size of Image 2) Shorter set-timeout. Can you have suggest any other?

@smfr - Any suggestion / help? Or can I close this PR?

**Comment by Ahmad-S792** [comment]
> > @smfr - I took test from Chromium patch and tried to apply two tricks to make it less flaky fo future: 1) Small Size of Image 2) Shorter set-timeout. Can you have suggest any other?
> 
> @smfr - Any suggestion / help? Or can I close this PR?

@smfr - I am closing my PR. If you want me to reopen, happy to do so and do any changes, I will keep my branch for time being. I have uploaded video on bug as well that it does indeed fix bug.

File: LayoutTests/svg/repaint/resources/animate-center.svg [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+       #square {
+         fill: red;
+         animation: keyframes 0.20s;
+         animation-fill-mode: forwards;
+         animation-timing-function: step-end;
+       }
+       @keyframes keyframes {
+         0% { fill: pink }
+         100% { fill: green }
+       }
```
Comment (graouts) [inline LayoutTests/svg/repaint/resources/animate-center.svg:13]: If you use the same value from `from` and `to`, and provided it's different to the underlying value (not `red`), does this help with the flakiness? You'd also need to ensure that this modified version of the test also fails prior to your patch.

**Pull request closed without merging**